src/udx.c

```diff
@@ -1614,312 +1794,333 @@ udx_socket_send_ttl (udx_socket_send_t *req, udx_socket_t *handle, const uv_buf_
 
-    if (pkt->type == UDX_PACKET_SEND) {
-      udx_socket_send_t *req = pkt->ctx;
+  while (socket->send_queue.len > 0) {
+    udx_packet_t *pkt = udx__fifo_shift(&socket->send_queue);
+    assert(pkt != NULL);
+    assert(pkt->type == UDX_PACKET_TYPE_SOCKET_SEND);
 
-      if (req->on_send != NULL) {
-        req->on_send(req, UV_ECANCELED);
-      }
-
-      continue;
+    udx_socket_send_t *req = pkt->ctx;
+    if (req->on_send) {
+      req->on_send(req, UV_ECANCELED);
     }
-
-    // stream packet, allow them to flush, by requeueing them
-    // flips the order but these are all state packets so whatevs
-
-    pkt->send_queue = &handle->send_queue;
-    pkt->fifo_gc = udx__fifo_push(&handle->send_queue, pkt);
   }
 
-  if (handle->send_queue.len == 0) {
-    udx__close_handles(handle);
+  if (!check_if_streams_have_data(socket)) {
+    udx__close_handles(socket);
   }
 
   return 0;
 }
 
 int
-udx_stream_init (udx_t *udx, udx_stream_t *handle, uint32_t local_id, udx_stream_close_cb close_cb) {
+udx_stream_init (udx_t *udx, udx_stream_t *stream, uint32_t local_id, udx_stream_close_cb close_cb) {
   ref_inc(udx);
 
-  handle->local_id = local_id;
-  handle->remote_id = 0;
-  handle->set_id = 0;
-  handle->status = 0;
-  handle->out_of_order = 0;
-  handle->recovery = 0;
-  handle->socket = NULL;
-  handle->relayed = false;
-  handle->relay_to = NULL;
-  handle->udx = udx;
-
-  handle->reordering_seen = false;
-  handle->retransmitting = 0;
-
-  handle->hit_high_watermark = false;
-  handle->writes_queued_bytes = 0;
-
-  handle->remote_changing = false;
-  handle->on_remote_changed = NULL;
-  handle->seq_on_remote_changed = 0;
-
-  handle->mtu = UDX_MTU_BASE;
-  handle->mtu_state = UDX_MTU_STATE_BASE;
-  handle->mtu_probe_wanted = false;
-  handle->mtu_probe_count = 0;
-  handle->mtu_probe_size = UDX_MTU_BASE; // starts with first ack, counts as a confirmation of base
-  handle->mtu_max = UDX_MTU_MAX;         // revised in connect()
-
-  handle->seq = 0;
-  handle->ack = 0;
-  handle->remote_acked = 0;
-
-  handle->srtt = 0;
-  handle->rttvar = 0;
-  handle->rto = 1000;
-  handle->rto_timeout = get_milliseconds() + handle->rto;
-  handle->rack_timeout = 0;
-
-  handle->rack_rtt_min = 0;
-  handle->rack_rtt = 0;
-  handle->rack_time_sent = 0;
-  handle->rack_next_seq = 0;
-  handle->rack_fack = 0;
-
-  handle->deferred_ack = 0;
-
-  handle->pkts_waiting = 0;
-  handle->pkts_inflight = 0;
-  handle->pkts_buffered = 0;
-  handle->retransmits_waiting = 0;
-  handle->seq_flushed = 0;
-
-  handle->sacks = 0;
-  handle->inflight = 0;
-  handle->ssthresh = 255;
-  handle->cwnd = UDX_CONG_INIT_CWND;
-  handle->cwnd_cnt = 0;
-  handle->rwnd = 0;
-
-  handle->on_firewall = NULL;
-  handle->on_read = NULL;
-  handle->on_recv = NULL;
-  handle->on_drain = NULL;
-  handle->on_close = close_cb;
+  stream->local_id = local_id;
+  stream->remote_id = 0;
+  stream->set_id = 0;
+  stream->status = 0;
+  stream->write_wanted = 0;
```

Comment:
what does this one indicate?

Reply:
it's a bitfield that indicates a need to write either a state packet or a destroy packet (needing to write a data packet is already indicated by having write_queue.len > 0 or retransmit_queue.len > 0). the difference vs the previous code is it sets this flag instead of queuing a packet immediately.